### PR TITLE
Rest-framework task

### PR DESCRIPTION
**Pull request by mansoorahmed786**

### CODE REVIEW

File: task2/task2/app/admin.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    add_fieldsets = (
+        (
+            None,
+            {
+                "classes": ("wide",),
+                "fields": ("email", "password1", "password2"),
+            },
+        ),
+    )
+admin.site.register(Profile)
```
Comment (talhamasood0000) [inline task2/task2/app/admin.py:58]: list_display
list_filter

File: task2/task2/app/serializers.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        return user
+    
+class ProjectSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Project
+        # fields = '__all__'
+        exclude = ['team_members']
+
+    def validate(self,data):
+        print("valisate")
```
Comment (talhamasood0000) [inline task2/task2/app/serializers.py]: remove this

File: task2/task2/app/models.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    is_active = models.BooleanField(default=True)
+    is_staff = models.BooleanField(default=False)
+    date_joined = models.DateTimeField(default=timezone.now)
+    objects = UserManager()
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name']
+    def save(self, *args, **kwargs):
+        super(User, self).save(*args, **kwargs)
+        return self
+
```
Comment (talhamasood0000) [inline task2/task2/app/models.py:87]:
```
class AllChocies:
    Add variables and values
```

File: task2/task2/app/serializers.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        return data
+
+class TaskSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Task
+        fields = '__all__'
+
+        def validate(self, data):
+            status = data.get('status')
+            if status not in ["open","review","working","awaiting_release","waiting_qa"]:
```
Comment (talhamasood0000) [inline task2/task2/app/serializers.py]: Change according to the model

File: task2/task2/app/tests.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        self.assertEqual(self.profile.user, self.user)
+        self.assertEqual(self.profile.role, "manager")
+
+class ProjectModelTestCase(TestCase):
+    def setUp(self):
+        project = Project.objects.create(title="Test Project", description="Test Description", start_date="2023-01-01", end_date="2023-02-01")
+    def test_project_creation(self):
+        self.assertEqual(self.project.title, "Test Project")
+        self.assertEqual(self.project.description, "Test Description")
+
```
Comment (talhamasood0000) [inline task2/task2/app/tests.py:33]: Spaces issue

File: task2/task2/app/views.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+        return Response(status=status.HTTP_201_CREATED,data=request.data)
+
+    def list(self, request, *args, **kwargs):
+        user = request.user
+        queryset = Task.objects.filter(assignee=user)
+        serializer = self.get_serializer(queryset, many=True)
+        return Response(serializer.data)
+    
+    def retrieve(self, request, *args, **kwargs):
+        instance = self.get_object()
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: missing condition

File: task2/task2/app/views.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    def custom_action(self, request, *args, **kwargs):
+        user = request.user
+        role =user.profile.role
+        if role!='manager':
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        partial = kwargs.pop('partial', True)
+        instance = self.get_object()
+        serializer = self.get_serializer(instance, data=request.data, partial=partial)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: Update project 
add validations

File: task2/task2/app/views.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+            instance = self.get_object()
+            instance.delete()
+        else:
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+    
+
+    @action(detail=True, methods=['PUT'])
+    def custom_action(self, request, *args, **kwargs):
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: new api to remove an assignee

File: task2/task2/app/views.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+###########################################################
+class DocumentViewSet(ModelViewSet):
+    serializer_class = DocumentSelializer
+    permission_classes = [IsAuthenticated]
+    queryset = Document.objects.all()
+
+    def create(self,request,*args,**kwargs):
+        project_id = request.data.get('project')
+        print(project_id)
+        project = Project.objects.filter(id=project_id).first()
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: validations missing. Move to serializer

File: task2/task2/app/views.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        if tasks.assignee != request.user :
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return Response(status=status.HTTP_201_CREATED,data=request.data)
+
+    def list(self, request, *args, **kwargs):
+        user = request.user
+        project = Project.objects.filter(team_members=user).first()
```
Comment (talhamasood0000) [inline task2/task2/app/views.py:194]: validations missing. move to serializer

File: task2/task2/app/views.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        queryset = Document.objects.filter(project=project)
+        serializer = self.get_serializer(queryset, many=True)
+        return Response(serializer.data)
+    
+    def retrieve(self, request, *args, **kwargs):
```
Comment (talhamasood0000) [inline task2/task2/app/views.py:199]: validations missing

File: task2/task2/app/views.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        return Response(status=status.HTTP_204_NO_CONTENT)
+    
+################################################################
+class CommentViewSet(ModelViewSet):
+    serializer_class = CommentSelializer
+    permission_classes = [IsAuthenticated]
+    queryset = Comment.objects.all()
+
+    def create(self,request,*args,**kwargs):
+        serializer = self.get_serializer(data=request.data)
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: validations missing

File: task2/task2/app/views.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        serializer.save()
+        return Response(status=status.HTTP_200_OK,data=request.data)
+    
+    def delete(self, request, *args, **kwargs):
+        user = request.user
+        rolee =user.profile.role
+        if rolee == 'manager':
+            instance = self.get_object()
+            instance.delete()
+        elif user==rolee:
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: chhange logic

File: task2/task2/app/views.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+        else:
+            queryset = Project.objects.filter(team_members=user)
+        serializer = self.get_serializer(queryset, many=True)
+        return Response(serializer.data)
+    
+    def retrieve(self, request, *args, **kwargs):
+        instance = self.get_object()
+        serializer = self.get_serializer(instance)
+        return Response(serializer.data)
+
```
Comment (talhamasood0000) [inline task2/task2/app/views.py:64]: add action to add a user in a team
detail true /project/{project_id}/add_member

request_data = {"user_id":1}

File: task2/task2/app/admin.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            {
+                "classes": ("wide",),
+                "fields": ("email", "password1", "password2"),
+            },
+        ),
+    )
+
+
+class TaskAdmin(admin.ModelAdmin):
+    list_display = ('title', 'status', 'project', 'assignee')
```
Comment (talhamasood0000) [inline task2/task2/app/admin.py]: list_display = ('__str__', 'status', 'project', 'assignee')

File: task2/task2/app/views.py [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+    
+class ProjectViewSet(ModelViewSet):
+    serializer_class = ProjectSerializer
+    permission_classes = [IsAuthenticated]
+    queryset = Project.objects.all()
+
+    def create(self, request, *args, **kwargs):
+        user = request.user
+        role =user.profile.role
+        if role!='manager':
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: ProjectChoices.MANAGER

File: task2/task2/app/views.py (hunk 14 above)
Comment (talhamasood0000) [reply task2/task2/app/views.py:64]: still missing

File: task2/task2/app/views.py [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+
+    
+    def delete(self, request, *args, **kwargs):
+        user = request.user
+        rolee =user.profile.role
+        if rolee == 'manager':
+            instance = self.get_object()
+            instance.delete()
+        elif user == rolee:
+            instance = self.get_object()
```
Comment (talhamasood0000) [inline task2/task2/app/views.py:70]: user cannot delete task

File: task2/task2/app/views.py [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+        role = user.profile.role
+        if role != "manager":
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return Response(status=status.HTTP_201_CREATED,data=request.data)
+
+    def list(self, request, *args, **kwargs):
+        user = request.user
```
Comment (talhamasood0000) [inline task2/task2/app/views.py:109]: missing for manager

File: task2/task2/app/views.py [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+        serializer = self.get_serializer(queryset, many=True)
+        return Response(serializer.data)
+    
+    def retrieve(self, request, *args, **kwargs):
+        user = request.user
+        rolee = user.profile.role
+        assignee = Task.objects.filter(assignee=user)
+        if rolee!='manager':
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        elif user == assignee:
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: error here

File: task2/task2/app/views.py [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+        serializer.save()
+        return Response(status=status.HTTP_200_OK,data=request.data)
+    
+    def delete(self, request, *args, **kwargs):
+        user = request.user
+        rolee =user.profile.role
+        if rolee!='manager':
+            instance = self.get_object()
+            instance.delete()
+        elif user==rolee:
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: remove this elif

File: task2/task2/app/views.py [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+        user = request.user
+        role =user.profile.role
+        if role!='manager':
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        
+        instance = self.get_object()
+        your_field_value = request.data.get('assignee')
+        instance.assignee = your_field_value
+        instance.save()
+        serializer = self.get_serializer(instance)
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: extra serializer

File: task2/task2/app/views.py [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+    @action(detail=True, methods=['DELETE'])
+    def delete_assignee(self, request, *args, **kwargs):
+        user = request.user
+        role =user.profile.role
+        if role!='manager':
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        
+        instance = self.get_object()
+        instance.delete()
+        return Response(status=status.HTTP_200_OK)
```
Comment (talhamasood0000) [inline task2/task2/app/views.py]: 204 no_content

File: task2/task2/app/views.py (hunk 9 above)
Comment (talhamasood0000) [reply task2/task2/app/views.py]: still missing

File: task2/task2/app/views.py (hunk 16 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py]: done

File: task2/task2/app/views.py (hunk 17 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py:70]: done

File: task2/task2/app/views.py (hunk 18 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py:109]: done

File: task2/task2/app/views.py (hunk 19 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py]: done

File: task2/task2/app/views.py (hunk 20 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py]: done

File: task2/task2/app/views.py (hunk 22 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py]: done

File: task2/task2/app/views.py (hunk 14 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py:64]: done

File: task2/task2/app/views.py (hunk 21 above)
Comment (mansoorahmed786) [reply task2/task2/app/views.py]: done

File: task2/task2/app/views.py [hunk 23]
```diff
@@ -0,0 +1,285 @@
...
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+        return Response(status=status.HTTP_201_CREATED,data=request.data)
+
+    def update(self, request, *args, **kwargs):
+        user = request.user
+        role =user.profile.role
+        if role!=ProjectChoices.MANAGER:
+            return Response(status=status.HTTP_403_FORBIDDEN,data={"details":"you dont have permission to access!"})
+        partial = kwargs.pop('partial', False)
```
Comment (talhamasood0000) [inline task2/task2/app/views.py:43]: partial=True
To update any single field

**Pull request merged**